Create documentation for espnet2 (#1710)
espnet/espnet: doc/espnet2_tutorial.md

# ESPnet2
We are planning super major update, called `ESPnet2`.

The developing status is still **under construction** yet, so please be very careful to  use with understanding following cautions:

- There might be fatal bugs related to essential parts.
- We haven't achieved comparable results to espnet1 on each tasks yet.

## Main changing from ESPnet1

- **Chainer free**
  - Discarding Chainer completely.
  - The development of Chainer is stopped at v7: https://chainer.org/announcement/2019/12/05/released-v7.html
- **Kaldi free**
  - It's not mandatory to compile Kaldi.
      - **If you find some recipes requiring Kaldi mandatory, please report it. It should be dealt as a bug in ESPnet2.**
  - We still support the features made by Kaldi optionally.
  - We still follow Kaldi style. i.e. depending on `utils/` of Kaldi.
- **On the fly** feature extraction & text preprocessing for training

>   - You don't need to create the features file before training, but just inputting wave data directly.

Reviewer: ```suggestion
  - You don't need to create the feature file before training, but just input wave data directly.
```

>       - We support both raw wave inputting and extracted feature.

Reviewer: ```suggestion
      - We support both raw wave input and extracted feature.
```

>   - The preprocessing for Text, tokenization to characters, or sentencepieces, can be also applied in training time.

Reviewer: ```suggestion
  - The preprocessing for text, tokenization to characters, or sentencepieces, can be also applied during training.
```

> - Discarding JSON format for describing training corpus in espnet1.

Reviewer: ```suggestion
- Discarding the JSON format for describing the training corpus in espnet1.
```

>     - Why do we discard JSON format? Because a dict object generated from large JSON file requires much memory and it also takes much times to parse such large JSON file.

Reviewer: ```suggestion
    - Why do we discard the JSON format? Because a dict object generated from a large JSON file requires much memory and it also takes much time to parse such a large JSON file.
```

- Support distributed data-parallel training (Not enough tested)
   - Single node multi GPU training with `DistributedDataParallel` is also supported.

## Recipes using ESPnet2

You can find the new recipes in `egs2`:

```
espnet/   # Python modules of epsnet1
espnet2/  # Python modules of epsnet2
egs/      # espnet1 recipes
egs2/     # espnet2 recipes
```

The usage of recipes is **almost same** as that of ESPnet1.

1. Change directory to the base directory

    ```bash
    # e.g.
    cd egs2/an4/asr1/
    ```

    `an4` is tiny corpus and can be freely obtained, so it might be suitable for this tutorial.

    Note that all shell scripts are supposed to be executed in the same level directory, `egs2/*/asr1`, and they can't be used in other place.

    ```bash
    # Doesn't work
    cd egs2/an4/
    ./asr1/run.sh
    ```

    ```bash
    # Doesn't work
    cd egs2/an4/asr1/local/
    ./data.sh
    ```

> 

Reviewer: It's better to add the case that it does work.

1. Change the configuration

    ```
    egs2/an4/asr1/
      - conf/      # Configuration files for training, inference, etc.
      - scripts/   # Bash utilities of espnet2
      - pyscripts/ # Python utilities of espnet2
      - steps/     # Kaldi utilities
      - utils/     # Kaldi utilities
      - db.sh      # The directory path of each corpora
      - path.sh    # Setup script for environment variables
      - cmd.sh     # Configuration for your backend of job scheduler
      - run.sh     # Entry point
      - asr.sh     # Invoked by run.sh
    ```

>     - You need to modify `db.sh` for specifying your corpus before executing `run.sh`. For example, you'll the recipe of `egs2/wsj`, you need to change the value of `WSJ0` and `WSJ1`.

Reviewer: ```suggestion
    - You need to modify `db.sh` for specifying your corpus before executing `run.sh`. For example, when you' touch the recipe of `egs2/wsj`, you need to change the paths of `WSJ0` and `WSJ1` in `db.sh`.
```

>         - Some corpus can be freely obtained from WEB and they are written as "downloads/" at the initial state. You can also change them to your corpus path if it's already downloaded.

Reviewer: ```suggestion
        - Some corpora can be freely obtained from the WEB and they are written as "downloads/" at the initial state. You can also change them to your corpus path if it's already downloaded.
```

>     - `path.sh` is used to setup the environment for `run.sh`. Note that the Python interpreter used for ESPnet is not the current Python of your terminal, but it's the Python which was installed at `tools/`. Thus you need to source `path.sh` to use the Python.

Reviewer: ```suggestion
    - `path.sh` is used to set up the environment for `run.sh`. Note that the Python interpreter used for ESPnet is not the current Python of your terminal, but it's the Python which was installed at `tools/`. Thus you need to source `path.sh` to use this Python.
```

    ```bash
    source path.sh
    python
    ```
    - `cmd.sh` is used for specifying the backend of JOB scheduler. If you don't have such system in your machine environment, you don't need to change anything about this file. See [Using Job scheduling system](./parallelization.md)

1. Run `run.sh`

    ```bash
    ./run.sh
    ```

    `run.sh` is an example script, which we often call as "recipe", to run all stages related to DNN experiments; data-preparation, training, and evaluation.

## See training status

### Log file

```bash
less exp/asr_train_<some-name>/train.log
```

### Show images

```bash
# Accuracy plot
# (eog is Eye of GNOME Image Viewer)
eog exp/asr_train_<some-name>/images/acc.img
# Attention plot
eog exp/asr_train_<some-name>/<sample-id>/<param-name>.img
```

### Use tensorboard

```bash
tensorboard --logdir exp/asr_train_<some-name>/tensorboard/
```

# Instruction for run.sh

> We uses all python commands via `run.sh` and you may need to invoke Python script directly in some cases, but `run.sh` itself is configurable and you can live comfortably with it in many cases without any modifications.

Reviewer: ```suggestion
We use all python commands via `run.sh` and you may need to invoke a Python script directly in some cases, but `run.sh` itself is configurable and you can live comfortably with it in many cases without any modifications.
```

## How to parse command line arguments in shell scripts?

All shell script in espnet/espnet2 depends on [utils/parse_options.sh](https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/utils/parse_options.sh) to parase command line arguments.

e.g. If the script has `ngpu` option

```bash
#!/bin/bash
# run.sh
ngpu=1
. utils/parse_options.sh
echo ${ngpu}
```

Then you can change the value as following:

```bash
$ ./run.sh --ngpu 2
echo 2
```

> You can also show the help message if it supports.

Reviewer: ```suggestion
You can also show the help message whether it is supported.
```
???

```bash
./run.sh --help
```

## Start from specified stage and stop at specified stage
The procedures in `run.sh` can be divided into some stages, e.g. data preparation, training, and evaluation. You can specify the starting stage and the stopping stage.

```bash
./run.sh --stage 2 --stop-stage 6
```

## Change the configuration for training
See  about the usage of training tool.

> (The following is the case of ASR training and you need to replace accordingly)

Reviewer: ```suggestion
(The following is the case of ASR training and you need to replace it accordingly)
```

```bash

> # Give configuration file

Reviewer: ```suggestion
# Give a configuration file
```

     - scripts/   # Bash utilities of espnet2
     - pyscripts/ # Python utilities of espnet2
     - steps/     # From Kaldi utilities
     - utils/     # From Kaldi utilities
     - db.sh      # The directory path of each corpora
     - path.sh    # Setup script for environment variables
     - cmd.sh     # Configuration for your backend of job scheduler
     - run.sh     # Entry point
     - asr.sh     # Invoked by run.sh
    ```

    - You need to modify `db.sh` for specifying your corpus before executing `run.sh`. For example, when you touch the recipe of `egs2/wsj`, you need to change the paths of `WSJ0` and `WSJ1` in `db.sh`.
      - Some corpora can be freely obtained from the WEB and they are written as "downloads/" at the initial state. You can also change them to your corpus path if it's already downloaded.
    - `path.sh` is used to set up the environment for `run.sh`. Note that the Python interpreter used for ESPnet is not the current Python of your terminal, but it's the Python which was installed at `tools/`. Thus you need to source `path.sh` to use this Python.
        ```bash
        source path.sh
        python
        ```

>     - `cmd.sh` is used for specifying the backend of the job scheduler. If you don't have such a system in your machine environment, you don't need to change anything about this file. See [Using Job scheduling system](./parallelization.md)

Reviewer: ```suggestion
    - `cmd.sh` is used for specifying the backend of the job scheduler. If you don't have such a system in your local machine environment, you don't need to change anything about this file. See [Using Job scheduling system](./parallelization.md)
```

1. Run `run.sh`

    ```bash
    ./run.sh
    ```

    `run.sh` is an example script, which we often call as "recipe", to run all stages related to DNN experiments; data-preparation, training, and evaluation.

## See training status

### Show the log file

```bash
% tail -f exp/asr_train_<some-name>/train.log
[host] 2020-04-05 16:34:54,278 (trainer:192) INFO: 2/40epoch started. Estimated time to finish: 7 minutes and 58.63 seconds
[host] 2020-04-05 16:34:56,315 (trainer:453) INFO: 2epoch:train:1-10batch: iter_time=0.006, forward_time=0.076, loss=50.873, los
s_att=35.801, loss_ctc=65.945, acc=0.471, backward_time=0.072, optim_step_time=0.006, lr_0=1.000, train_time=0.203
[host] 2020-04-05 16:34:58,046 (trainer:453) INFO: 2epoch:train:11-20batch: iter_time=4.280e-05, forward_time=0.068, loss=44.369
, loss_att=28.776, loss_ctc=59.962, acc=0.506, backward_time=0.055, optim_step_time=0.006, lr_0=1.000, train_time=0.173
```

### Show the training status in a image file

```bash
# Accuracy plot
# (eog is Eye of GNOME Image Viewer)
eog exp/asr_train_<some-name>/images/acc.img
# Attention plot
eog exp/asr_train_<some-name>/att_ws/<sample-id>/<param-name>.img
```

> ### Use tensorboard

Reviewer: Maybe better to add some descriptions about how to install tensorboard and a simple usage like https://espnet.github.io/espnet/tutorial.html#logging.
We don't include the tensorboard installation as default (we could do it though).

Reviewer: Maybe it is OK to leave it as it is but in the near future, we may need to prepare a complete README for espnet2 that merges both espnet1 and espnet2 documentations.

Author: Sorry, I already added tensorboard.

https://github.com/espnet/espnet/blob/eca4fa073a826957c91069501844d4e2b9392cc9/setup.py#L33-L34

This is  because pytorch1.1? officially start to support tensorboard logging and it requires to install tensorboard. 
I decided to use the pytorch's SummaryWriter instead of tensorboardx if pytorch>=1.1, so we need to install tensorboard.

Reviewer: Make sense for me.
This is very good.
